addons/mail/static/src/models/thread/thread.js

```diff
@@ -190,6 +200,48 @@ function factory(dependencies) {
                     ];
                 }
             }
+            if ('seen_partners_info' in data) {
+                if (!data.seen_partners_info) {
+                    data2.partnerSeenInfos = [['unlink-all']];
+                } else {
+                    data2.partnerSeenInfos = [
+                        ['insert-and-replace',
+                            data.seen_partners_info.map(
+                                ({ fetched_message_id, id, partner_id, seen_message_id}) => {
+                                    return {
+                                        id,
+                                        lastFetchedMessage: [seen_message_id ? ['insert', {id: seen_message_id}] : ['unlink-all']],
+                                        lastSeenMessage: [fetched_message_id ? ['insert', {id: fetched_message_id}] : ['unlink-all']],
+                                        partner: [['insert', {id: partner_id}]],
+                                    };
+                                })
+                        ]
+                    ];
+                    if (data.id || this.id) {
+                        const messageIds = data.seen_partners_info.reduce((currentSet, { fetched_message_id, seen_message_id}) => {
+                            if (fetched_message_id) {
+                                currentSet.add(fetched_message_id);
+                            }
+                            if (seen_message_id) {
+                                currentSet.add(seen_message_id);
+                            }
+                            return currentSet;
+                        }, new Set());
+                        if (messageIds.size > 0) {
+                            data2.messageSeenIndicators = [
+                                ['insert',
+                                    [...messageIds].map(messageId => {
+                                       return {
+                                           id: this.env.models['mail.message_seen_indicator'].computeId(messageId, data.id || this.id),
+                                           message: [['insert', {id: messageId}]],
+                                       };
+                                    })
+                                ]
+                            ];
+                        }
```

Comment:
This would work fine without "set" because insert is not creating duplicate anyway, and relational fields guarantee uniqueness of linked records.

It is slightly more efficient like this to not do insert many times with the same value, but I'm not sure if it's worth the more complex code.

Reply:
Mostly FYI, probably not worth a change considering time available.